[Merged by Bors] - feat(data/equiv/derangements/basic): define derangements

HenrySwanson opened the pull request:
This proves two formulas for the number of derangements on _n_ elements, and defines some combinatorial equivalences
involving derangements on α and derangements on certain subsets of α. This proves Theorem 88 on Freek's list.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

This is rather messy; would really appreciate suggestions. I read through a few mathlib files, but I'm definitely not
super-familiar with Lean conventions yet.

I have a bunch of specific things I'd like feedback on, with `-- TODO` comments.

I'm also trying to get something with `equiv.perm.decompose_option` working, but it's not looking promising. :| There's a nice equivalence from `perm (option α)` to `(option α) x (perm α)` that could be used. It restricts on derangements to `α x {f : perm α // [no fixed points, or f(none) is the only fixed point]}`, which is tempting, for obvious reasons. But the intermediate casting of subtypes is hard. I'll update this if it works out, but it's not looking like it'll be any cleaner.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

HenrySwanson committed:
feat(archive/100-theorems-list): add proof of derangement formula

This is rather messy; would really appreciate suggestions. I read
through a few mathlib files, but I'm definitely not super-familiar
with Lean conventions yet.

HenrySwanson committed:
Fix header message, and update 100.yaml

oops, left the wrong name in after copy-pasting

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
```
First comment --- absolutely everything in this PR belongs in mathlib proper. No need to add anything to `archive/100-theorems-list/` except a pointer to the result.

This will require moving things to a bunch of different files!

HenrySwanson replied:
would it go somewhere like `combinatorics/derangements.lean` and `data/equiv/<something>.lean`? (depending on if it's about perms or derangements specifically)

awainverse replied:
I imagine that the default location for things moved from this should be to a new file `data/equiv/perm/derangements.lean`

kim-em replied:
@HenrySwanson, when a conversation has been dealt with, it's easiest if the author takes the initiative to click "Resolve conversation". I think this one is good now?

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  let a_and_b : set α := {a, b},
+
+  -- TODO why can't it infer this is decidable?
+  haveI : decidable_pred a_and_b := by
```
Type inference doesn't do any unfolding of definitions, so I think it just can't see through the `let`. Is there any reason to have the `let` in the first place?

HenrySwanson replied:
without it i have to specify `{a, b} : set \alpha`, and sometimes i have to wrap that in a `( ... : Type _)` to make the checker happy. i'm not sure why this is :\

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+    conv_rhs { rw apply_eq_iff_eq_symm_apply, rw apply_eq_iff_eq_symm_apply },
+    simp [swap, or_comm] }),
+
+  transitivity {f : perm α // ∀ x : a_and_b, f x = swap' x},
```
Surely you want `x` mem `a_and_b`, not to coerce `a_and_b` to a type here?

HenrySwanson replied:
if i do that, then `equiv.set.compl` doesn't get the input it expects. not sure if there's a counterpart that takes sets and \mem

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  let swap' : perm a_and_b := perm.subtype_perm swap (by {
+    intro x,
+
+    simp,
```
No non-terminal `simp`: use `simp?` or `squeeze_simp` and replace this with a `simp only [...]`.

HenrySwanson replied:
oh interesting! didn't know that was a general guideline. is that for performance reasons?

jcommelin replied:
It improves performance a bit, but it's mostly for maintenance reasons. If someone adds a new simp-lemma to mathlib, suddenly simp might do more (or different) to what it used to do, and the rest of the proof will break. So then someone (who typically doesn't really know this proof) has to come and fix it. And so then you have to go back to an old copy of mathlib, to find what the goal after the `simp` is supposed to look like. You can see why that's annoying.
With `simp only [lem1, lem2, ...]` it's basically guaranteed to always find the same result. (Unless someone changes the statement of `lem1`, but then that person probably knows quite easily how to fix this proof, because it is exactly clear which ingredients `simp` is using (and not using).

HenrySwanson replied:
oh interesting, that makes sense! so i should convert all my non-terminal `simp`s to `simp only`s?

ericrbg replied:
yep! also, you can go for `suffices [TACTIC STATE AFTER SIMP], by simpa`, which I quite like

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+end
+
+-- TODO should this be a simp lemma?
+lemma perm_ignore_two_cycle_eq (a b : α) (f : perm α) (hf : swaps f a b) (c : ({a, b}ᶜ : set α)):
```
Again, use set membership rather than coercing sets to types, where possible.

HenrySwanson replied:
the equiv library sometimes puts up a fight about that, but i'll convert as much as i can

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  ((perm_ignore_swap a b) ⟨f, hf⟩ c : α) = f c := by simp [perm_ignore_swap, equiv.set.compl]
+
+-- TODO this name definitely isn't up to snuff... but what would it be called?
+lemma perm_ignore_two_cycle_derangement_iff (a b : α) (hab : a ≠ b) (f : perm α) (hf : swaps f a b) :
```
`is_derangement_iff_ignore_swap`?

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  is_derangement f ↔ is_derangement ((perm_ignore_swap a b) ⟨f, hf⟩) :=
+begin
+  split,
+
```
Usually don't leave blank spaces in proof environments. It's okay, but even better is to fill that blank space with a helpful comment. :-)

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+
+-- TODO: these are just here to help the elaborator. I would like to eliminate
+-- them in favor of `:` coercions, but I keep getting errors about it. :\
+def everything_but {α : Type*} (a : α) : set α := {a}ᶜ
```
What's wrong with just writing out `({a}ᶜ : set α)` each time? It's about the same length anyway.

HenrySwanson replied:
i had some issues with not being able to infer the right typeclass (something something `has_coe_to something`). lemme see if i can make it happen again

ericrbg replied:
Actually, I've had issues with \compl too before; I sometimes can't write it in front of a set and have to write `.compl`, which is defeq. Henry, does it work if you use `.compl`?

HenrySwanson replied:
whoop, forgot to resolve this, but no, i got the same issue with "can't infer typeclass" :(

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+section counting
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
```
I'm not sure it's even worth introducing this. How does it look to just write it out each time?

HenrySwanson replied:
IMO it's like the catalan numbers, where you can just say `a(n)` without reference to a specific set of n elements

not super necessary but i think it's nice

ericrbg commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+    rw finset.card_insert_of_not_mem,
+    { simp },
+    { change _ ≠ _ at hm,
+      simp [hm.symm] }
```
```suggestion
    simp [ne.symm hm]
```

kim-em commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+end
+
+-- TODO `n!` inside or outside the sum? should n!/k! be a function outputting ℕ?
+theorem num_derangements_sum (n : ℕ) :
```
There's another PR #7363 that introduces `desc_fac`, which is just what you need here. Perhaps you can either copy that over, or coordinate pulling that out into a separate PR we can merge quickly.

ericrbg replied:
I'll move it out into another PR Scott ^^

ericrbg replied:
#7527

HenrySwanson replied:
yeah, that'd be great :o the fact that we have to go into Z here is unavoidable, but staying away from Q is a possibility

ericrbg commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  -- TODO dunno why this didn't work out with simp, but whatever
+  simp [fintype.card] at card_everything_but,
+  rw [key, card_everything_but],
+  ring,
```
```suggestion
  rw key,
  simpa [finset.card_univ, card_everything_but] using mul_add _ _ _,
```

HenrySwanson replied:
what's the difference between the `using` clause, and just putting it inside the simp?

ericrbg replied:
it basically does `simp` and then `exact`s what's in the `using` clause when it can; putting inside the `simp` could work too! i didn't do it because I worried it could loop (if something brackets everything back up again) but it probably works just fine

Ruben-VandeVelde commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  refine subtype_equiv (perm_congr e) _,
+  intro f,
+  unfold derangements is_derangement,
+  simp,
```
```suggestion
  simp only [derangements, is_derangement, perm_congr_apply, set.mem_set_of_eq],
```

Ruben-VandeVelde commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+      use [x, x_ne_a],
+      simp [subtype.ext_iff_val, perm_splice_out_eq],
+      rw apply_eq_iff_eq_symm_apply,
+      simp [x_fixed, swap_apply_of_ne_of_ne x_ne_a x_ne_b] } },
```
```suggestion
    rintro (rfl|⟨x, x_fixed⟩),
    { use [b, hab.symm],
      simp [subtype.ext_iff_val, perm_splice_out_eq] },
    { have x_ne_a : x ≠ a,
      { rintro rfl,
        rw x_fixed at fa_eq_b,
        contradiction },

      have x_ne_b : x ≠ b,
      { rintro rfl,
        refine hab (f.injective _),
        rw [fa_eq_b, x_fixed] },

      use [x, x_ne_a],
      simp [subtype.ext_iff_val, perm_splice_out_eq, x_fixed,
        swap_apply_of_ne_of_ne x_ne_a x_ne_b] } },
```

Ruben-VandeVelde commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,468 @@
+  have hab : a ≠ b,
+  { rintro rfl,
+    simp [everything_but] at hab,
+    exact hab },
```
```suggestion
    simpa [everything_but] using hab },
```

Ruben-VandeVelde reviewed (commented):
Some stylistic suggestions:

HenrySwanson committed:
Apply suggestions from code review

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>
Co-authored-by: Eric <37984851+ericrbg@users.noreply.github.com>

HenrySwanson committed:
Update with option-based approach

It turns out that working with `derangements (option \alpha)` was much
easier, mostly because working with `\alpha` is easier than working with
the set `{a}^c` for some `a`.

So this got around a lot of the issues with sets/types and giving things
names to appease the elaborator. There's still a few more TODOs, but
I've incorporated the changes that were suggested (e.g. no non-terminal
simps, style guidelines, pull things into files).

awainverse commented on archive/100-theorems-list/88_derangements_formula.lean:
```diff
@@ -0,0 +1,451 @@
+-- TODO feels like this should already exist? something something equiv_functor?
+-- I tried mul_action.to_perm (perm α) (perm α), but that didn't simp easily
+/-- Post-multiplying by a given `f : perm α` gives a permutation on `perm α`. -/
+def equiv_compose {α : Type*} (f : perm α) : perm (perm α) :=
```
This seems like it should at least work for all groups (as right-multiplication), even if `mul_action.to_perm` doesn't work for some reason. Could you point out a few places where `simp` gets mad at `mul_action.to_perm`, so I could try to help troubleshoot that?

HenrySwanson replied:
Sure! Here's what happens when I use `mul_action.to_perm` in place of defining my own. There's a lot of definitions I have to expand before it arrives at "just compose the permutations": https://pastebin.com/32ZhFC55

HenrySwanson committed:
Oops, missed something in the refactor

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
```
```suggestion
def is_derangement {α : Type*} (f : perm α) : Prop := ∀ x : α, f x ≠ x
```

eric-wieser replied:
Using `support` this becomes
```suggestion
def is_derangement {α : Type*} (f : perm α) : Prop := f.support = set.univ
```

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
+
+/-- The permutation `f` has exactly one fixed point, `a`. -/
+def exactly_one_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x ↔ x = a
```
These shouldn't be in the global namespace. Perhaps put them inside `perm`, and use dot notation?

HenrySwanson replied:
sure sure, can do

eric-wieser replied:
Are these statements easier to make in terms of `perm.support`?

HenrySwanson replied:
Oh, I didn't know that existed! I'll give that a try

eric-wieser replied:
The first would be `{a}ᶜ ⊆ f.support` and the second `{a}ᶜ = f.support`. These seem attractive as they avoid the need for either of these definitions.

eric-wieser replied:
Ah, this needs finiteness. @pechersky, you had some work on this in review, right?

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+section TODO_name
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangement_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
```
```suggestion
def derangements_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
```

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+    { rintro ⟨rfl⟩, exact fa_eq_a }}
+end
+
+lemma derangement_iff_opfp_and_ne (a : α) (f : perm α) :
```
```suggestion
lemma is_derangement_iff_opfp_and_ne (a : α) (f : perm α) :
```

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+
+-- TODO I defined this predicate `foo` and the corresponding set, but it's only
+-- used in the proof of `derangements_equiv_sigma_opfp`. Should it be completely
+-- folded into that proof?
```
`foo_lemma` is clearly much too long to inline.

HenrySwanson replied:
yeah, guess i'll cook up a name of some kind. probably should wrap it behind a namespace too

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+
+-- TODO this kinda makes more sense as a subset of `α × perm α`, but the proof was easier this way.
+/-- The set of derangements on `option α` is equivalent to the union of "permutations with either
+    no fixed points, `a` as the only fixed point", over all `a : α`. -/
```
```suggestion
    no fixed points or `a` as the only fixed point", over all `a : α`. -/
```

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+  derangements (option α) ≃ Σ a : α, {f : perm α // only_possible_fixed_point f a} :=
+begin
+  -- TODO why do i need the @ to get it to infer α?
+  let fiber : option α → set (perm α) := λ a : option α, {f : perm α | (a, f) ∈ (@foo_set α)},
```
Maybe just make the argument to `foo_set` explicit, anyway?

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+  have fiber_equiv_opfp : ∀ a : α, (fiber a) ≃ {f : perm α // only_possible_fixed_point f a},
+  { intro a, refl },
+
+  -- TODO how are calc blocks this big formatted?
```
This is fine. Certainly align the equality symbols, as you've done. It's nice to align the `:`s, but it's okay to break this rule to save lines, as you've done.

eric-wieser commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,265 @@
+    rintro ⟨x, hx_fixed⟩,
+    use e x,
+    simp [hx_fixed] }
+end
```
```suggestion
subtype_equiv (perm_congr e) $ λ f, e.forall_congr $ λ x, by simp
```

HenrySwanson committed:
Naming changes

HenrySwanson committed:
Use desc_fac

HenrySwanson committed:
Merge branch 'master' into HEAD

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+begin
+  intro f,
+  apply fintype.decidable_forall_fintype,
+end
```
```suggestion
instance : decidable_pred (@is_derangement α) := λ _, fintype.decidable_forall_fintype
```

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+begin
+  have : fintype (perm α) := by apply_instance,
+  dsimp [derangements],
+  exact set_fintype (set_of is_derangement),
```
```suggestion
  rw [derangements],
  apply_instance
```

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+  card (derangements α) = num_derangements (card α) :=
+begin
+  unfold num_derangements,
+  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
```
```suggestion
  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
```

most of the time, if you nudge Lean enough you don't need to unfold much

ericrbg replied:
also, if you chose this because `rw` wasn't cooperating; `erw` can sometimes deal with stuff like this (and it does in this case)

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+    simp },
+  have key := card_congr (@derangements_recursion_equiv X _),
+  rw [num_derangements_invariant, fintype.card_option, fintype.card_fin] at key,
+  simp [card_everything_but, mul_add, key],
```
```suggestion
  have card_everything_but : ∀ a : fin (n+1), card ({a}ᶜ : set (fin (n+1))) = n,
  { intro a,
    simp only [fintype.card_of_finset, set.mem_compl_singleton_iff],
    rw [finset.filter_ne' _ a, finset.card_erase_of_mem (finset.mem_univ a)],
    simp },
  have key := card_congr (@derangements_recursion_equiv (fin (n+1)) _),
  rw [num_derangements_invariant, fintype.card_option, fintype.card_fin] at key,
  simp [card_everything_but, mul_add, key]
```

there's a lot of `singleton`s so it got confused which one you wanted, + also it was overeagerly `coe`ing to `Sort` so you were trying to `compl` a `Sort`.

ericrbg replied:
I do think there really should be a better way to do stuff in the middle... I think there should be a theorem saying that if you have `s : set α` and `[fintype s]`, then `card s = s.to_finset.card`; the `finset` card API could do a lot better with this than the mess you had to do

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+  use derangements_congr (equiv_fin α),
+end
+
+theorem num_derangements_recursive (n : ℕ) :
```
maybe this needs a better name... `num_derangements_succ`?

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+  push_cast,
+  -- TODO can these proofs be inferred from some tactic? i tried linarith,
+  -- but for some reason that didn't work
+  rw hyp n (nat.lt_succ_of_le (nat.le_succ _)),
```
```suggestion
  rw hyp n (nat.lt_succ_of_le (nat.le_succ _)),
```

`omega` can do these but its use is strongly discouraged right now as it's not getting ported to Lean 4; it's probably just worth doing them by hand

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+    -- simplify the arguments to desc_fac, and then evaluate them explicitly
+    simp only [nat.sub_self, nat.add_sub_cancel_left, nat.desc_fac_succ, nat.desc_fac_zero],
+    push_cast,
+    ring },
```
```suggestion
  { -- simplify the arguments to desc_fac, and then evaluate them explicitly
    norm_num [pow_succ, ←nat.add_one, nat.desc_fac_succ, nat.desc_fac_zero],
    ring },
```

maybe this is `norm_num` abuse but it works fine :b

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+  -- show that (n+1) (-1)^k / (n! / k!) + (n+1) (-1)^k ((n+1)! / k!) = (-1)^k ((n+2)! / k!)
+  { refine finset.sum_congr rfl _,
+    -- introduce k and rewrite the constraint to `k ≤ n`
+    intros k h_mem,
```
I wonder if this bit there's a better way to do this by cases on `k`; the (-1)^k terms are constant throughout so they really don't matter, and then you can work in ℕ which may be easier

HenrySwanson replied:
i could try multiplying through by (-1)^k; that should be easier than removing it with division

HenrySwanson replied:
that didn't seem to change the difficulty any; the fussing with n occurs inside the argument of `num_derangements`, which remains un-zify-able

ericrbg commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,134 @@
+    rw add_assoc },
+  -- now both sides are of the form `sum (x : ℕ) in finset.range(n+1) (...) + (...)`, in such
+  -- a way that we can just check that the LHS and RHS match up in each summand
+  congr' 1,
```
i feel like this is doable without `conv`? maybe with a `simp only [sum_range_succ]` and then just tidy up a little bit?

HenrySwanson replied:
i tried that once, early on, but i got a lot more extra terms (since we end up with finset.range(n) at the end instead of finset.range(n+1)), and the constant terms got a bit unwieldy. maybe with desc_fac they'll work out though

HenrySwanson replied:
i rearranged things a bit so that the conv blocks are smaller, but because we need to arrange terms in a particular order for `congr'` to work out, i couldn't get rid of them entirely without causing a mess

ericrbg commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,225 @@
+  is_derangement f ↔ only_possible_fixed_point f a ∧ f a ≠ a :=
+begin
+  unfold is_derangement only_possible_fixed_point,
+  split,
```
```suggestion
  split,
```

ericrbg commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,225 @@
+    intro f,
+    simp only [exists_prop, set.mem_set_of_eq],
+    rw and_comm,
+    exact eofp_iff_opfp_and_eq _ _ },
```
```suggestion
    simp_rw [exists_prop, set.mem_set_of_eq, eofp_iff_opfp_and_eq, and_comm] },
```

ericrbg reviewed (commented):
some stuff I saw; I didn't look at the `option` section at all

HenrySwanson committed:
Apply suggestions from code review

Co-authored-by: Eric <37984851+ericrbg@users.noreply.github.com>

HenrySwanson committed:
Clean up use of conv

Getting rid of it entirely is really messy, but this seems to strike a
good balance.

Also renamed num_derangements_recursive to num_derangements_succ.

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,222 @@
+    simp_rw (eofp_iff_opfp_and_eq a),
+    exact subtype_subtype_equiv_subtype_inter _ (λ f : perm α, f a = a),
+    apply_instance,
+  },
```
Brackets don't need their own lines.

kim-em commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,222 @@
+
+variables {α : Type*} [decidable_eq α]
+
+def remove_none.fiber (a : option α) : set (perm α) :=
```
Still missing doc-strings. Can you check the linter?

kim-em commented on docs/100.yaml:
```diff
@@ -327,6 +327,10 @@
+    decls  :
+    - num_derangements_recursive
+    - num_derangements_sum
+  author : Henry Swanson
```
Check the linter: something is broken here.

ericrbg replied:
```suggestion
88:
  title  : Derangements Formula
  decls  :
    - num_derangements_recursive
    - num_derangements_sum
  author : Henry Swanson
```

i think YAML cares about indents

HenrySwanson committed:
Incorporate suggestions

Man I am struggle-bussing to get this PR in shape here.

kim-em reviewed (approved):
This is looking pretty good! Sorry it has taken a while to get through review.

I'm happy to hit merge soon if no one else chimes in (or merges first).

ericrbg commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,223 @@
+  remove_none F = f :=
+begin
+  simp [remove_none.fiber, derangements],
+end
```
```suggestion
  remove_none F = f := by simp [remove_none.fiber, derangements],
```

ericrbg reviewed (approved):
apart from this, I'd be tempted to agree Scott!

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,223 @@
+def is_derangement {α : Type*} (f : perm α) : Prop := ∀ x : α, f x ≠ x
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
```
`is_derangement` and `derangements` are quite literally the same thing. A term of type `set (perm α)` is the same as a function `perm α → Prop`. And you've defined them so that they have equal values. I suggest you keep only one of them, depending on whether you care about set notation or not (I think in this particular case you do).

YaelDillies commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,117 @@
+  use derangements_congr (equiv_fin α),
+end
+
+theorem num_derangements_succ (n : ℕ) :
```
I think it would be nice to use this inductive relationship as a definition and then to show `num_derangements_invariant` from there. That shouldn't change too much the proofs, it gets you rid of this weird reference to `fin n`, and allows for better evaluation (I think?).

YaelDillies replied:
After reflexion, a better inductive definition might be
```
def num_derangements : ℕ → ℕ
| 0       := 1
| (n + 1) := (n + 1) * num_derangements n + (-1 : ℤ)^n
```
(it doesn't work as of because -1 isn't a natural, but that shouldn't be too hard to overcome, maybe by splitting cases on parity)

HenrySwanson replied:
It's rather tricky to go about that -- the reason the proof for `num_derangements_succ` goes through is that we can ignore the complicated specifics of the type (`fin(n+1)` minus a particular element), and instead just consider the cardinality.

I tried something with `induction h_card : (card α) generalizing α`, and it's a lot less clear. Maybe I'll just replace `num_derangements` with `num_derangements_aux` and then show that your definition matches the original.

YaelDillies reviewed (changes requested):
This looks nice. I do not mean to discourage you, but for me the interesting part of the derangement formula is that `n!/(num_derangements n)` tends to `e`.

eric-wieser commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,223 @@
+/-- The set of permutations `f` such that the preimage of `(a, f)` under
+    `perm.decompose_option` is a derangement. -/
+def remove_none.fiber (a : option α) : set (perm α) :=
+  {f : perm α | (a, f) ∈ equiv.perm.decompose_option '' derangements (option α)}
```
This might be easier to work with as:
```suggestion
  {f : perm α | (equiv.perm.decompose_option.symm (a, f)).is_derangement}
```

HenrySwanson replied:
This setup makes it match `equiv.set_prod_equiv_sigma`, so I think it's easier to work with in the original form

eric-wieser commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,117 @@
+
+lemma num_derangements_0 : num_derangements 0 = 1 := rfl
+
+lemma num_derangements_1 : num_derangements 1 = 0 := rfl
```
This lemma and the one above should be `simp`

HenrySwanson replied:
Oddly enough, `simp [num_derangements]` times out for me here :(

Not entirely sure why simp doesn't know these are "defeq" (is that the term?)

ericrbg replied:
I think Eric meant `@[simp]`

HenrySwanson replied:
Oh! Yeah now I get it

HenrySwanson committed:
Incorporate suggestions

Notably, split up the "size of derangements on fin(n)" and "recursive
definition" definitions.

eric-wieser commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+def remove_none.fiber (a : option α) : set (perm α) :=
+  {f : perm α | (a, f) ∈ equiv.perm.decompose_option '' derangements (option α)}
+
+lemma remove_none.fiber_mem (a : option α) (f : perm α) :
```
```suggestion
lemma remove_none.mem_fiber (a : option α) (f : perm α) :
```

HenrySwanson committed:
Attempt to prove the limit is 1/e

I can work with tendsto on its own, and I think I could work with
cau_seq, but together, I'm stuck.

HenrySwanson committed:
Prove that derangements(n)/n! -> 1/e

https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/cau_seq.20and.20tendsto

Only things left are bookkeeping -- where do these lemmas go, etc.

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+import data.equiv.option
+import group_theory.perm.option
+
+import tactic.equiv_rw
```
```suggestion
import group_theory.perm.option
import tactic.equiv_rw
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+    split,
+    { exact h_opfp x },
+    { rintro ⟨rfl⟩, exact fa_eq_a }}
+end
```
```suggestion
⟨λ h, ⟨λ x, (h x).mp, (h a).mpr rfl⟩, λ h x, ⟨h.1 x, forall_eq.mpr h.2 x⟩⟩
```

HenrySwanson replied:
When is it better to write out the explicit proof term, and when is it better to leave it human readable? Is the former better here because it's pretty mechanically uninteresting?

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+  { rintros ⟨h_opfp, fa_ne_a⟩ x x_fixed,
+    specialize h_opfp x x_fixed,
+    rw h_opfp at x_fixed,
+    exact fa_ne_a x_fixed }
```
```suggestion
  refine ⟨λ h_derangement, ⟨λ x x_fixed, (h_derangement x x_fixed).elim, h_derangement a⟩, _⟩,
  rintros ⟨h_opfp, fa_ne_a⟩ x x_fixed,
  specialize h_opfp x x_fixed,
  rw h_opfp at x_fixed,
  exact fa_ne_a x_fixed
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+  let key_equiv : {f : perm α | f a = a} ≃ perm ({a}ᶜ : set α),
+  { transitivity {f : perm α // ∀ x : ({a} : set α), f x = (equiv.refl _) x},
+    { apply subtype_equiv_right, simp },
+    { exact equiv.set.compl (equiv.refl _) }},
```
```suggestion
    { exact equiv.set.compl (equiv.refl _) } },
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+      cases x,
+      { simp },
+      { simp only [equiv_functor.map_equiv_apply, equiv_functor.map,
+                    option.map_eq_map,  option.map_some'],
```
```suggestion
                    option.map_eq_map, option.map_some'],
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+      { simp only [equiv_functor.map_equiv_apply, equiv_functor.map,
+                    option.map_eq_map,  option.map_some'],
+        by_cases x_vs_a : x = a,
+        { rw [x_vs_a, swap_apply_right], exact option.some_ne_none _, },
```
```suggestion
        { rw [x_vs_a, swap_apply_right], apply option.some_ne_none },
```
You don't seem to include a comma inside the brace anywhere else.

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+          intro contra,
+          exact x_vs_a (h_opfp x contra) }},
+    },
+    { rw apply_symm_apply }}
```
```suggestion
          exact x_vs_a (h_opfp x contra) } } },
    { rw apply_symm_apply } }
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+begin
+  have fiber_is_some : ∀ a : option α, (remove_none.fiber a) → a.is_some,
+  { intro a,
+    cases a,
```
```suggestion
  { rintro ⟨a⟩,
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+  { intro a,
+    cases a,
+    { rw remove_none.fiber_none_eq_empty, simp },
+    { intro _, simp }},
```
```suggestion
    { intro _, simp } },
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+    simp [is_derangement_iff_opfp_and_ne a] },
+
+  -- there's some decidable_pred goals generated by equiv_rw; this clears
+  -- them out
```
```suggestion
  -- there are some decidable_pred goals generated by equiv_rw; this clears them out
```

benjamindavidson commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,218 @@
+    apply_fun some at x_fixed,
+    cases Fx : F (some x) with y,
+    { rw remove_none_none F Fx at x_fixed, rw ←option.some_inj, rwa x_fixed at F_none },
+    { exfalso, rw remove_none_some F ⟨y, Fx⟩ at x_fixed, exact F_derangement _ x_fixed }},
```
```suggestion
    { exfalso, rw remove_none_some F ⟨y, Fx⟩ at x_fixed, exact F_derangement _ x_fixed } },
```

benjamindavidson commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,122 @@
+    congr' 1,
+    -- show that (n + 1) * (-1)^x * desc_fac x (n - x) = (-1)^x * desc_fac x (n.succ - x)
+    {
+      refine finset.sum_congr (refl _) _,
```
```suggestion
    { refine finset.sum_congr (refl _) _,
```

benjamindavidson commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,122 @@
+      rw [nat.succ_sub h_le, nat.desc_fac_succ, nat.add_sub_cancel' h_le],
+      push_cast,
+      ring,
+    },
```
```suggestion
      ring },
```

benjamindavidson commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,122 @@
+      ring,
+    },
+    -- show that -(-1)^n = (-1)^n.succ * desc_fac n.succ (n.succ - n.succ)
+    { simp [pow_succ] }}
```
```suggestion
    { simp [pow_succ] } }
```

HenrySwanson committed:
Apply suggestions from code review

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

eric-wieser commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,120 @@
+
+/-- The number of derangements on an `n`-element set. This definition is bad for
+    computation though, use `num_derangements` instead. -/
+def num_derangements_aux (n : ℕ) : ℕ := card (derangements (fin n))
```
I'd be inclined to remove this definition entirely and inline it into the few places where it is used

HenrySwanson replied:
I've gotten rid of this now, and you're right, it's nicer.

HenrySwanson committed:
Remove unnecessary imports

HenrySwanson committed:
Merge branch 'derangements' of https://github.com/leanprover-community/mathlib into derangements

HenrySwanson committed:
Merge remote-tracking branch 'origin/master' into derangements

HenrySwanson committed:
Merge remote-tracking branch 'origin/master' into derangements

HenrySwanson committed:
Replace desc_fac with asc_factorial

Dang, easier than expected. Why did I put this off for so long again?

HenrySwanson committed:
Whoops, forgot some TODOs.

Let's just flag those as comments in GH.

HenrySwanson commented on src/data/equiv/derangements/exponential.lean:
```diff
@@ -0,0 +1,71 @@
+open_locale big_operators
+open_locale topological_space
+
+lemma complex.tendsto_iff_real (u : ℕ → ℝ) (x : ℝ) :
```
This must already exist, but I was unable to find it. Any hints?

YaelDillies replied:
Can't the proof be `⟨(complex.continuous_of_real.tendsto x).comp, (complex.continuous_re.tendsto x).comp⟩`? This `comp` suggests that your statement might not be general enough. But I don't know enough about mathlib analysis to tell for sure.

HenrySwanson commented on src/data/equiv/derangements/exponential.lean:
```diff
@@ -0,0 +1,71 @@
+begin
+  rw complex.tendsto_iff_real,
+  convert complex.tendsto_exp_series x; simp,
+end
```
These should probably go elsewhere. With the definition of exp, maybe? The cau_seq and tendsto incompatibilities mean I didn't quite know what to do here for a while.

YaelDillies replied:
You can't put them in `data.complex.exponential` because it's missing the instance `topological_space ℂ`... I don't know where to find it. I think it's being inferred.

HenrySwanson committed:
Fuss around with the particular connection between card (derangements)
and the nicely-computable num_derangements.

I think @eric-weiser was right after all. Also it gives me a chance to
fix the lint errors that weren't caught locally.

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+  exactly_one_fixed_point f a ↔ only_possible_fixed_point f a ∧ f a = a :=
+⟨λ h, ⟨λ x, (h x).mp, (h a).mpr rfl⟩, λ h x, ⟨h.1 x, forall_eq.mpr h.2 x⟩⟩
+
+lemma is_derangement_iff_opfp_and_ne (a : α) (f : perm α) :
```
```suggestion
lemma mem_derangements_iff_opfp_and_ne (a : α) (f : perm α) :
```

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+
+In this file we define a predicate and its corresponding set:
+
+* `is_derangement f`: whether the permutation `f` has no fixed points.
```
No need to keep this anymore. l.12 above also needs updating.

YaelDillies commented on src/data/equiv/derangements/exponential.lean:
```diff
@@ -0,0 +1,71 @@
+  let s : ℕ → ℝ := λ n, ∑ k in finset.range n, (-1 : ℝ)^k / k.factorial,
+  -- this isn't entirely obvious, since we have to ensure that desc_fac and factorial interact in
+  -- the right way, e.g. that k stays less than n
+  have : ∀ n : ℕ, (num_derangements n : ℝ) / n.factorial = s(n+1),
```
This could well be a `suffices`

YaelDillies commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,149 @@
+  ring,
+end
+
+lemma card_derangements_eq_num_derangements_fin {n : ℕ} :
```
```suggestion
lemma card_derangements_fin_eq_num_derangements {n : ℕ} :
```

YaelDillies commented on src/data/equiv/derangements/finite.lean:
```diff
@@ -0,0 +1,149 @@
+end
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements : ℕ → ℤ
```
Could you find a definition that goes `ℕ → ℕ`? This works (as you've shown with `num_derangements_succ_succ`):
```
def num_derangements : ℕ → ℕ
| 0 := 1
| 1 := 0
| (n + 2) := (n + 1) * (num_derangements (n + 1) + num_derangements n)
```
and by looking at your proof of `card_derangements_eq_num_derangements_fin` I think you only ever need this recurrence relation.

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+def derangements (α : Type*) : set (perm α) := {f : perm α | ∀ x : α, f x ≠ x}
+
+/-- The permutation `f` has at most one fixed point, `a`. -/
+def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
```
This definition and the next could be stated more generally than for `perm`. Maybe simply `function.only_possible_fixed_point` and `function.exactly_one_fixed_point`? They could go in `dynamics.fixed_points`.
And maybe you should reuse their definition of `fixed_points`, to which point you may not even need to define anything new.

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+    { apply subtype_equiv_right, simp },
+    { exact equiv.set.compl (equiv.refl _) } },
+
+  transitivity {f : {f : perm α // f a = a} // only_possible_fixed_point f.val a},
```
That could look nicer as a `calc` block, I don't know.

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+section simple_lemmas
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangements_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
```
I don't know for sure, but I think this should rather be called `equiv.derangements`, so outside of the `perm` namespace.

YaelDillies commented on src/data/equiv/derangements/basic.lean:
```diff
@@ -0,0 +1,201 @@
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | ∀ x : α, f x ≠ x}
```
This could be, importing `dynamics.fixed_points`,
```suggestion
def derangements (α : Type*) : set (perm α) := {f : perm α | fixed_points f = ∅}
```
or
```suggestion
def derangements (α : Type*) : set (perm α) := {f : perm α | fixed_points f ⊆ ∅}
```
But maybe it's better to have that as a lemma rather than as the def.

YaelDillies reviewed (commented):
Okay, this is definitely too much material for one PR. Sorry for having asked you for the exponential stuff here earlier.
You can probably PR `derangement.basic` as a standalone, then `derangement.finite`, then `derangement.exponential`.

I think you should reuse stuff about fixed points we already have in `dynamics.fixed_points`. Another file that uses them is `order.fixed_points`, if that sparks you ideas.

Putting all that stuff under `data.equiv` sounds odd. Could it go under `combinatorics`, rather?

HenrySwanson committed:
Did a bunch of small changes from PR

Fixed comments
Changed some function names
Simplified complex.tendsto_iff_real and num_derangements_tendsto_e
Changed num_derangements codomain to \N

HenrySwanson committed:
Move files to combinatorics/derangements

HenrySwanson committed:
Remove use of equiv_rw

It was really making a lot of messy extra goals. Refine is nicer.

HenrySwanson committed:
Merge remote-tracking branch 'origin/master' into derangements

HenrySwanson committed:
Add a "reduce option T -> T in sigma's left" lemma

This seems like it might be nice for combinatorics.

YaelDillies commented on src/combinatorics/derangements/finite.lean:
```diff
@@ -0,0 +1,147 @@
+    number of derangements.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
```
```suggestion
* `card_derangements_invariant`: A lemma stating that the number of derangements on a type `α`
  depends only on the cardinality of `α`.
* `num_derangements n`: The number of derangements on an n-element set, defined in a computation-
  friendly way.
 * `card_derangements_eq_num_derangements`: Proof that `num_derangements` really does compute the
  number of derangements.
* `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
  factorials.
```
so that docgen doesn't mess up with it

YaelDillies committed:
branch(combinatorics/derangements/*): use function.fixed_points (#8690)

* initial commit

* transparent perm.compl_equiv

* finish derangements.compl_equiv

* finish

HenrySwanson committed:
Fix lint error with simp lemma, and clean up remove_none.fiber_eq_opfp a
little more

HenrySwanson committed:
Erase finite.lean and exponential.lean

Those will be added in later PRs

HenrySwanson committed:
Remove links from 100.yaml

Linter was unhappy

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+
+lemma function.mem_fixed_points_iff {α : Type*} {f : α → α} {x : α} :
+  x ∈ function.fixed_points f ↔ f x = x :=
+by refl
```
```suggestion
iff.rfl
```

YaelDillies replied:
Have you tried it? 😛 `iff.rfl` isn't strong enough here.

ericrbg replied:
yeah, it had worked for me on my local

HenrySwanson replied:
`iff.rfl _` works for me. i'll update to that, but i'm curious why one would be preferred over the other

YaelDillies replied:
Oh okay, must've messed it up when I tried.

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+  left_inv := equiv.perm.subtype_perm_of_subtype,
+  right_inv := begin
+    rintro ⟨f, hf⟩,
+    exact subtype.ext (equiv.perm.of_subtype_subtype_perm _ $ λ a hfa ha, hfa $ hf _ ha),
```
```suggestion
  right_inv := λ ⟨f, hf⟩,
    subtype.ext (equiv.perm.of_subtype_subtype_perm _ $ λ a hfa ha, hfa $ hf _ ha) }
```

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+
+/-- Permutations on a subtype are equivalent to permutations on the original type that fix pointwise
+the rest. -/
+protected def perm.subtype_equiv {α : Type*} (p : α → Prop) [decidable_pred p] :
```
this is just the same as `compl_equiv`, right? (if you end up keeping this, fix the `right_inv` as above)

YaelDillies replied:
See #8691

ericrbg replied:
crap, didn't know there was another PR for this

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def equiv.derangements_congr (e : α ≃ β) : (derangements α ≃ derangements β) :=
+e.perm_congr.subtype_equiv $ λ f, e.forall_congr $ λ x, by simp
```
```suggestion
e.perm_congr.subtype_equiv $ λ f, e.forall_congr $ by simp
```

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+
+/-- Derangements on a subtype are equivalent to permutations on the original type whose set of fixed
+points is the rest. -/
+protected def subtype_equiv  (p : α → Prop) [decidable_pred p] :
```
```suggestion
protected def subtype_equiv (p : α → Prop) [decidable_pred p] :
```

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+        refine hf _ _ ha,
+        rw [mem_fixed_points, is_fixed_pt, perm.subtype_equiv],
+        dsimp,
+        rw [equiv.perm.of_subtype_apply_of_mem _, hfa, subtype.coe_mk],
```
this should probably be a lemma of `subtype_equiv`

HenrySwanson replied:
I didn't quite understand this comment, and I've revised this since then. @ericrbg does the newest revision address this?

ericrbg replied:
looks fine to me. i was suggesting refactoring part of this proof out, but the new one seems ok

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+/- Same as `(equiv.refl _)^.set.compl .symm.trans (subtype_equiv_right $ by simp)` but with better
+unfolding -/
+/-- Permutations on `sᶜ` are equivalent to permutations that fix `s` pointwise. -/
+protected def perm.compl_equiv {α : Type*} (s : set α) [decidable_pred (∈ s)] :
```
probably worth jamming `@[simps]` on all these defs and seeing if anything breaks or not

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+namespace derangements
+
+/-- Derangements on a subtype are equivalent to permutations on the original type whose set of fixed
+points is the rest. -/
```
```suggestion
/-- Derangements on a subtype are equivalent to permutations on the original type where points are
fixed iff they are not in the subtype. -/
```

Maybe this is more clear?

ericrbg commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,226 @@
+      : subtype_equiv_right (λ f, by simp_rw [exists_prop, ←forall_and_distrib,
+        ←iff_iff_implies_and_implies])
+
+/-- The set of permutations that fix at most `a` is equivalent to the sum of:
```
```suggestion
/-- The set of permutations that fix either `a` or nothing is equivalent to the sum of:
```

ericrbg reviewed (commented):
looks good to me, some random golfing/doc comments I saw

HenrySwanson committed:
Finalize things

Waited for #8768 and #8691 to land, and addressed remaining comments.
Things are looking up :)

HenrySwanson committed:
Merge remote-tracking branch 'origin/master' into derangements

HenrySwanson committed:
change some names

YaelDillies commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,190 @@
+
+/-- The set of derangements on `option α` is equivalent to the union over `a : α`
+    of "permutations with `a` the only possible fixed point". -/
+def derangements_equiv_sigma_at_most_one_fixed_point :
```
```suggestion
def derangements_option_equiv_sigma_at_most_one_fixed_point :
```

YaelDillies commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,190 @@
+namespace equiv
+variables [decidable_eq α]
+
+/-- The set of permutations `f` such that the preimage of `(a, f)` under
```
```suggestion
/-- The set of permutations `f` such that a preimage of `(a, f)` under
```
or is it really unique?

HenrySwanson replied:
it's unique once `a` is fixed, since `equiv.perm.decompose_option` is an equivalence

YaelDillies commented on src/data/equiv/basic.lean:
```diff
@@ -1426,6 +1426,21 @@ calc (Σ y : subtype q, {x : α // f x = y}) ≃
+  { intro x,
+    cases x,
+    { intro n, exfalso, exact h n },
+    { intro s, exact rfl }},
```
```suggestion
    { intro s, exact rfl } },
```

YaelDillies commented on src/data/equiv/basic.lean:
```diff
@@ -1426,6 +1426,21 @@ calc (Σ y : subtype q, {x : α // f x = y}) ≃
+    { intro n, exfalso, exact h n },
+    { intro s, exact rfl }},
+  exact equiv.trans
+    (sigma_subtype_equiv_of_subset _ _ h').symm
```
Slight golf
```suggestion
  exact (sigma_subtype_equiv_of_subset _ _ h').symm.trans
```

YaelDillies commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,190 @@
+          have ne_2 : some x ≠ some a := (option.some_injective α).ne_iff.mpr x_vs_a,
+          rw [swap_apply_of_ne_of_ne ne_1 ne_2, (option.some_injective α).ne_iff],
+          intro contra,
+          exact x_vs_a (h_opfp contra) } } },
```
```suggestion
      simp only [equiv_functor.map_equiv_apply, equiv_functor.map,
                  option.map_eq_map, option.map_some'],
      by_cases x_vs_a : x = a,
      { rw [x_vs_a, swap_apply_right], apply option.some_ne_none },
      have ne_1 : some x ≠ none := option.some_ne_none _,
      have ne_2 : some x ≠ some a := (option.some_injective α).ne_iff.mpr x_vs_a,
      rw [swap_apply_of_ne_of_ne ne_1 ne_2, (option.some_injective α).ne_iff],
      intro contra,
      exact x_vs_a (h_opfp contra) },
```
You don't need that much indenting

HenrySwanson replied:
if i've got two cases and the first is simple, i can just unindent the second then?

YaelDillies commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,190 @@
+* equivalences involving various subtypes of `perm α` and `derangements α`: e.g.,
+
+  - `derangements_equiv_sigma_opfp`: an equivalence between `derangements (option α)` and the
+    sigma-type `Σ a : α, {f : perm α // only_possible_fixed_point f a}`.
```
```suggestion
We also define some equivalences involving various subtypes of `perm α` and `derangements α`:
* `derangements_option_equiv_sigma_at_most_one_fixed_point`: An equivalence between
  `derangements (option α)` and the sigma-type `Σ a : α, {f : perm α // fixed_points f ⊆ a}`.
* `derangements_recursion_equiv`: An equivalence between `derangements (option α)` and the
  sigma-type `Σ a : α, (derangements (({a}ᶜ : set α) : Type _) ⊕ derangements α)` which is later
  used to inductively count the number of derangements.
```

YaelDillies commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,190 @@
+/-!
+# Derangements on types
+
+In this file we define `derangements α`, the set of derangements on a type `α`.
```
Maybe talk about `remove_none.fiber` somewhere?

YaelDillies reviewed (approved):
This is looking good!

Here's my final batch of suggestions. LGTM otherwise

HenrySwanson committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

HenrySwanson committed:
Add comment about remove_none.fiber

tb65536 commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,188 @@
+
+lemma mem_derangements_iff_fixed_points_eq_empty {f : perm α} :
+  f ∈ derangements α ↔ fixed_points f = ∅ :=
+set.eq_empty_iff_forall_not_mem.symm
```
You could also add a lemma `mem_derangements_iff_support_eq_top` (or something like that)

YaelDillies replied:
`function.support` is about `f x ≠ 0` and `function.mul_support` about `f x ≠ 1`. So that doesn't work. However I did think about having `derangements_on (s : set α) := {f : perm α | fixed_points f = sᶜ}` in which case your lemma suggestion could become `derangements α = derangements_on univ`.

HenrySwanson replied:
I thought about the `support` for permutations (on mobile so idk the exact file, but it was used in defining the sign of a permutation), but it was hard to use both that and `function.fixed_points`, so I picked just one. Perhaps the former could be revised to make use of the latter? Outside the scope of this PR though.

YaelDillies replied:
Well first I see that `equiv.perm.support` supposes `fintype α`, so it's enough for what you're doing here but underpowered otherwise.

HenrySwanson committed:
fix typo

kim-em commented on src/combinatorics/derangements/basic.lean:
```diff
@@ -0,0 +1,188 @@
+        { rw set.eq_empty_iff_forall_not_mem,
+          refine ⟨λ h x hx, h.2 (h.1 hx ▸ hx), λ h, ⟨λ x hx, (h _ hx).elim, h _⟩⟩ }
+      end
+  ... ≃ (derangements _) ⊕ derangements α
```
No harm in filling in the underscore here, for readability.

HenrySwanson committed:
Fill in underscore in calc block
